apps/glusterfs/volume_entry_allocate.go

```diff
@@ -63,6 +66,209 @@ func (v *VolumeEntry) allocBricksInCluster(db *bolt.DB,
 	}
 }
 
+func (v *VolumeEntry) getEntryfromBrickName(tx *bolt.Tx, brickname string) (brickEntry *BrickEntry, e error) {
+	var brickEntries []*BrickEntry
+	brickids := v.BricksIds()
+	for _, brickid := range brickids {
+		brick, err := NewBrickEntryFromId(tx, brickid)
+		if err != nil {
+			return nil, err
+		}
+		brickEntries = append(brickEntries, brick)
+	}
+
+	for _, brickentry := range brickEntries {
+		nodeEntry, err := NewNodeEntryFromId(tx, brickentry.Info.NodeId)
+		if err != nil {
+			return nil, err
+		}
+		if brickname == fmt.Sprintf("%v:%v", nodeEntry.Info.Hostnames.Storage[0], brickentry.Info.Path) {
+			return brickentry, nil
+		}
+	}
+	return nil, ErrNotFound
+}
+
+func (v *VolumeEntry) replaceBrickInVolume(db *bolt.DB, executor executors.Executor,
+	allocator Allocator,
+	oldBrickId string) (e error) {
+
+	logger.Info("Entered replace brick")
+	// Do the work in the database context so that the cluster
+	// data does not change while determining brick location
+	err := db.Update(func(tx *bolt.Tx) error {
+
+		oldBrickEntry, err := NewBrickEntryFromId(tx, oldBrickId)
+		if err != nil {
+			return err
+		}
+
+		oldDeviceEntry, err := NewDeviceEntryFromId(tx, oldBrickEntry.Info.DeviceId)
+		if err != nil {
+			return err
+		}
+		oldBrickNode, err := NewNodeEntryFromId(tx, oldBrickEntry.Info.NodeId)
+		if err != nil {
+			return err
+		}
+
+		// Determine the setlist by getting data from Gluster
+		vinfo, err := executor.VolumeInfo(oldBrickNode.ManageHostName(), v.Info.Name)
+		var slicestartindex int
+		var foundbrickset bool
+		var brick executors.Brick
+		setlist := make([]*BrickEntry, 0)
+		for slicestartindex = 0; slicestartindex <= len(vinfo.Bricks.Bricks)-v.Durability.BricksInSet(); slicestartindex = slicestartindex + v.Durability.BricksInSet() {
```

Comment:
I am hesitating to populate db with this much of info because I want those changes to be properly designed. I request this change be taken in in current form. As it isn't a API change, it can be done in subsequent patches without breaking backward compatibility.